src/libraries/System.Numerics.Tensors/src/System/Numerics/Tensors/TensorPrimitives.netcore.cs

```diff
@@ -348,9 +344,179 @@ private static float CosineSimilarityCore(ReadOnlySpan<float> x, ReadOnlySpan<fl
             return result;
         }
 
+        /// <remarks>
+        /// This is the same as <see cref="Aggregate{TLoad, TAggregate}(float, ReadOnlySpan{float})"/>,
+        /// except it early exits on NaN.
+        /// </remarks>
+        private static float MinMaxCore<TMinMax>(ReadOnlySpan<float> x) where TMinMax : struct, IBinaryOperator
+        {
+            if (x.IsEmpty)
+            {
+                ThrowHelper.ThrowArgument_SpansMustBeNonEmpty();
+            }
+
+            // This matches the IEEE 754:2019 `maximum`/`minimum` functions.
+            // It propagates NaN inputs back to the caller and
+            // otherwise returns the greater of the inputs.
+            // It treats +0 as greater than -0 as per the specification.
+
+            // Initialize the result to the identity value
+            float result = x[0];
+            int i = 0;
+
+#if NET8_0_OR_GREATER
+            if (Vector512.IsHardwareAccelerated && x.Length >= Vector512<float>.Count * 2)
+            {
+                ref float xRef = ref MemoryMarshal.GetReference(x);
+
+                // Load the first vector as the initial set of results, and bail immediately
+                // to scalar handling if it contains any NaNs (which don't compare equally to themselves).
```

Comment:
Why does this need to bail to the scalar path?

If we find `NaN`, we return `NaN`. We then have the option of returning the exact `NaN` (best practice) or a generic `NaN` (allowed by the IEEE 754 spec).

So we could either do:
```csharp
if (!Vector512.EqualsAll(resultVector, resultVector))
{
    return float.NaN;
}
```

or, something like this (which follows the "best practice"):
```csharp
if (!Vector512.EqualsAll(resultVector, resultVector))
{
    Vector512<float> vectorMask = Vector512.Equals(resultVector, resultVector);
    ulong bitMask = vectorMask.ExtractMostSignificantBits();

    int index = BitOperations.TrailingZeroCount(bitMask);
    return resultVector[index];
}
```

Reply:
It depends on the workload.

There are a lot of workloads that will avoid `NaN` entirely. There are other workloads that may use `NaN` as part of representing sparse or complex data.

I think it'd be fine to not handle it in the first vectorization pass here. But it'd probably be worth explicitly handling longer term.

Reply:
> I think it'd be fine to not handle it in the first vectorization pass here. But it'd probably be worth explicitly handling longer term.

If it's worth doing longer term, I'll just do it now. Why put off for tomorrow, and all that.